### ISSUE TITLE
Does compose still try to schedule all the containers to the same swarm node?

### ISSUE THREAD

**Issue by anentropic** [docker/compose#3042]
maybe I'm tripping but I thought this https://github.com/docker/swarm/issues/251 was going to/had been be fixed by the new overlay networking support

I've gone through a round of upgrading everything to the latest docker versions

the `mem_limit` in my compose yml works now, but as a result I can't start my entire set of containers... due to large amount being (correctly) allocated for one of them

```
$ docker info
Containers: 8
Images: 14
Role: primary
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 2
 oceania-1: ****:2376
  └ Status: Healthy
  └ Containers: 1
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 1.026 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.16.0-4-amd64, operatingsystem=Debian GNU/Linux 8 (jessie), provider=digitalocean, storagedriver=aufs
 oceania-master: ****:2376
  └ Status: Healthy
  └ Containers: 7
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 1 GiB / 1.026 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.16.0-4-amd64, operatingsystem=Debian GNU/Linux 8 (jessie), provider=digitalocean, storagedriver=aufs
CPUs: 2
Total Memory: 2.052 GiB
Name: 0dc444daa840
```

```
$ docker-compose -f docker-compose/live.yml -p oceania up -d
nginx is up-to-date
oceania_redis_1 is up-to-date
oceania_elasticsearch_1 is up-to-date
Creating oceania_myapp_1
ERROR: no resources available to schedule container
```

there's a big allocation for the elasticsearch, then it gets stuck trying to start the next one instead of scheduling it to the empty node

I have:
```
version: 2

networks:
  oceania:
    driver: overlay

services:
  nginx:
    image: nginx
    container_name: nginx
    networks:
      - oceania

  redis:
    image: redis
    networks:
      - oceania
    expose:
      - 6379
    restart: unless-stopped

  elasticsearch:
    image: elasticsearch:1.7.4
    networks:
      - oceania
    expose:
      - 9200
      - 9300
    mem_limit: 1G
    restart: unless-stopped

etc
```

Am I trying to do something that's not possible yet?  Or something is broken in my swarm?

**Comment by anentropic** [comment]
```
$ docker-machine ls
NAME             ACTIVE      DRIVER         STATE     URL               SWARM                     DOCKER    ERRORS
consul           -           digitalocean   Running   tcp://****:2376                             v1.10.2
oceania-1        -           digitalocean   Running   tcp://****:2376   oceania-master            v1.10.2
oceania-master   * (swarm)   digitalocean   Running   tcp://****:2376   oceania-master (master)   v1.10.2
```

**Comment by dnephin** [comment]
Everything looks correct from what I can see. 

Looking at the swarm source, this error is raised when there are no nodes that have the available cpu and memory constraints. Could it be that the other service has a `mem_limit` > 1G ?

**Comment by anentropic** [comment]
Hi, thanks very much for responding

I have two containers with `mem_limit`: `1G` for elasticsearch and `640M` for my web app

So it baulked at scheduling the web app after starting elasticsearch... but so far the `oceania-1` node is completely empty... it seems to be trying to schedule everything on the master

My web app has:
```
    links:
      - elasticsearch
```

so this is the heart of my question... I know in the past this forced compose to schedule them on the same node, but I thought it was to be fixed with overlay networking, and I'm wondering does it work yet?

**Comment by dnephin** [comment]
I see, that part of the Compose files was missing from the example above.

https://docs.docker.com/compose/swarm/ covers this.

`links` are still considered dependencies by swarm (they shouldn't be for overlay networks, that fix is being worked on in swarm). You can remove them from your Compose file, they aren't necessary anymore. If you want to define an order you can use `depends_on`.

**Comment by anentropic** [comment]
thanks very much for your help

I see now I don't need `links` with overlay network.  I do want to ensure that the elasticsearch is running before starting the web app service so I just changed the `links` to `depends_on`.  There are no `volumes_from` dependencies between the services.

unfortunately, same result:

```
$ docker-compose -f docker-compose/live.yml -p oceania stop
Stopping oceania_elasticsearch_1 ... done
Stopping oceania_redis_1 ... done
Stopping nginx ... done

$ docker-compose -f docker-compose/live.yml -p oceania up -d
Starting nginx
Starting oceania_redis_1
Starting oceania_elasticsearch_1
Creating oceania_myapp_1
ERROR: no resources available to schedule container
```

it has scheduled everything on the master again and then got stuck

**Comment by dnephin** [comment]
If you are running Compose 1.6.0 you might be hitting the issue fixed by #2894 in Compose 1.6.2. After upgrading only containers with unnamed volumes will be recreated on the same host.

You can also `docker-compose rm` to wipe out the old containers to confirm.

**Comment by anentropic** [comment]
thank you again!

I could have sworn blind I had upgraded docker-compose but a quick check and sure enough still on 1.6.0-rc2

after upgrading to 1.6.2 it works as intended!
